src/protocols/LoRaWAN/LoRaWAN.cpp

```diff
@@ -681,23 +835,63 @@ int16_t LoRaWANNode::downlink(uint8_t* data, size_t* len) {
   if(state == RADIOLIB_ERR_LORA_HEADER_DAMAGED) {
     state = RADIOLIB_ERR_NONE;
   }
+  
+  if(state != RADIOLIB_ERR_NONE) {
+    #if !defined(RADIOLIB_STATIC_ONLY)
+      delete[] downlinkMsg;
+    #endif
+    return(state);
+  }
 
   // get the frame counter and set it to the MIC calculation block
-  // TODO this will not handle overflow into 32-bits!
   // TODO cache the ADR bit?
-  uint16_t fcnt = LoRaWANNode::ntoh<uint16_t>(&downlinkMsg[RADIOLIB_LORAWAN_FHDR_FCNT_POS]);
-  LoRaWANNode::hton<uint16_t>(&downlinkMsg[RADIOLIB_LORAWAN_BLOCK_FCNT_POS], fcnt);
-
+  uint16_t fcnt16 = LoRaWANNode::ntoh<uint16_t>(&downlinkMsg[RADIOLIB_LORAWAN_FHDR_FCNT_POS]);
+  LoRaWANNode::hton<uint16_t>(&downlinkMsg[RADIOLIB_LORAWAN_BLOCK_FCNT_POS], fcnt16);
+  uint32_t fcnt32 = fcnt16;   // calculate possible rollover once decided if this is network downlink or application downlink
+  
   RADIOLIB_DEBUG_PRINTLN("downlinkMsg:");
   RADIOLIB_DEBUG_HEXDUMP(downlinkMsg, RADIOLIB_AES128_BLOCK_SIZE + downlinkMsgLen);
-  
-  if(state != RADIOLIB_ERR_NONE) {
+
+  // calculate length of FOpts and payload
+  uint8_t foptsLen = downlinkMsg[RADIOLIB_LORAWAN_FHDR_FCTRL_POS] & RADIOLIB_LORAWAN_FHDR_FOPTS_LEN_MASK;
+  int payLen = downlinkMsgLen - 8 - foptsLen - sizeof(uint32_t);
+
+  bool isAppDownlink = true;
+  if (payLen <= 0 && this->rev == 1) {  // no payload => MAC commands only => Network frame (LoRaWAN v1.1 only)
+    isAppDownlink = false;
+  }
+
+  // check the FcntDown value (Network or Application)
+  uint32_t fcntDownPrev = 0;
+  if (isAppDownlink) {
+    fcntDownPrev = mod->hal->getPersistentParameter<uint32_t>(RADIOLIB_PERSISTENT_PARAM_LORAWAN_A_FCNT_DOWN_ID);
+  } else {
+    fcntDownPrev = mod->hal->getPersistentParameter<uint32_t>(RADIOLIB_PERSISTENT_PARAM_LORAWAN_N_FCNT_DOWN_ID);
+  }
+
+  // assume a 16-bit to 32-bit rollover when difference in LSB is smaller than MAX_FCNT_GAP
+  // if that isn't the case and the received fcnt is smaller or equal to the last heard fcnt, then error
+  if (fcnt16 <= fcntDownPrev && 0xFFFF - (uint16_t)fcntDownPrev + fcnt16 > RADIOLIB_LORAWAN_MAX_FCNT_GAP) {
```

Comment:
Please add parentheses to make the order of operations explicit (this can lead to warning or error for some compiler configurations).